File: service/makeaplan/service/phase_service.go
package service

import (
	"context"
	"fmt"

	"github.com/guregu/null"
	"google.golang.org/grpc"

	pb "makeaplan.com/api/model"
	"makeaplan.com/api/service"
	"makeaplan.com/app"
	"makeaplan.com/dao"
	"makeaplan.com/model"
)

type phaseService struct {
}

func RegisterPhaseService(server *grpc.Server) {
	service.RegisterPhaseQueryService(server, service.NewPhaseQueryService(&phaseService{}))
	service.RegisterPhaseModifyService(server, service.NewPhaseModifyService(&phaseService{}))
}

func (p *phaseService) QueryPhaseSummary(ctx context.Context, r *service.QueryPhaseSummaryRequest) (*service.QueryPhaseSummaryResponse, error) {
	phase, err := dao.GetPhase(ctx, uint32(r.GetPhaseId()))
	if err != nil {
		return nil, err
	}

	ts, err := app.TaskStatis(uint32(r.GetPlanId()), &phase.ID)
	if err != nil {
		return nil, err
	}
	ps := &pb.PhaseSummary{
		Phase:      phase.AsPbPhase(),
		Status:     app.TimeBasedStatus(phase.StartTime, phase.EndTime),
		TaskStatis: ts,
	}

	return &service.QueryPhaseSummaryResponse{Phase: ps}, nil
}

func (p *phaseService) QueryPlanPhaseSummary(ctx context.Context, r *service.QueryPlanPhaseSummaryRequest) (*service.QueryPlanPhaseSummaryResponse, error) {
	ps, err := dao.GetPhaseByPlanID(r.GetPlanId())
	if err != nil {
		return nil, err
	}

	pss := make([]*pb.PhaseSummary, 0, len(ps))
	for _, p := range ps {
		ts, err := app.TaskStatis(uint32(r.GetPlanId()), &p.ID)
		if err != nil {
			return nil, err
		}
		pss = append(pss, &pb.PhaseSummary{
			Phase:      p.AsPbPhase(),
			Status:     app.TimeBasedStatus(p.StartTime, p.EndTime),
			TaskStatis: ts,
		})
	}

	return &service.QueryPlanPhaseSummaryResponse{Phases: pss}, nil
}

func (p *phaseService) UpdatePhase(ctx context.Context, r *service.UpdatePhaseRequest) (*service.UpdatePhaseResponse, error) {
	mp := &model.Phase{}
	mp.Title = r.GetTitle()
	if !app.IsBlankString(r.GetNote()) {
		mp.Note = null.StringFrom(r.GetNote().GetValue())
	}
	mp.EndTime = null.TimeFrom(r.GetEndTime().AsTime())
	if r.GetStartTime() != nil {
		mp.StartTime = null.TimeFrom(r.GetStartTime().AsTime())
	}
	err := dao.UpdatePhaseData(ctx, uint32(r.GetPhaseId()), mp)
	if err != nil {
		return nil, err
	}

	return &service.UpdatePhaseResponse{}, nil
}

func (p *phaseService) CreatePhase(ctx context.Context, r *service.CreatePhaseRequest) (*service.CreatePhaseResponse, error) {
	mp := &model.Phase{}
	mp.Title = r.GetTitle()
	if !app.IsBlankString(r.GetNote()) {
		mp.Note = null.StringFrom(r.GetNote().GetValue())
	}
	mp.EndTime = null.TimeFrom(r.GetEndTime().AsTime())
	if r.GetStartTime() != nil {
		mp.StartTime = null.TimeFrom(r.GetStartTime().AsTime())
	}

	newPhaseID, err := dao.AddPhaseInTransaction(ctx, mp, r.GetPlanId())
	if err != nil {
		return nil, fmt.Errorf("failed to dao.AddPhaseInTransaction, err: %v", err)
	}

	return &service.CreatePhaseResponse{PhaseId: int32(newPhaseID)}, nil
}

func (p *phaseService) DeletePhase(ctx context.Context, r *service.DeletePhaseRequest) (*service.DeletePhaseResponse, error) {
	mp := &model.Phase{}
	mp.PhaseStatus = model.PhaseStatusDeleted
	_, _, err := dao.UpdatePhase(ctx, uint32(r.GetPhaseId()), mp)
	if err != nil {
		return nil, fmt.Errorf("failed to dao.UpdatePhase, err: %v", err)
	}

	return &service.DeletePhaseResponse{}, nil
}

File: mobile/lib/service/phase_service.dart
import 'package:makeaplan/api/phase.pb.dart' as pb;
import 'package:makeaplan/api/phase_service.pb.dart' as pb;
import 'package:makeaplan/api/phase_service.pbgrpc.dart' as grpc;

import 'grpc_client.dart';

class PhaseService {
  static PhaseService _singleton;

  PhaseService._internal();

  factory PhaseService() {
    if (_singleton == null) {
      _singleton = PhaseService._internal();
    }
    return _singleton;
  }

  grpc.PhaseQueryClient _client() {
    return grpc.PhaseQueryClient(
      GRPCClient().client(),
      options: getCommonCallOptions(),
    );
  }

  grpc.PhaseModifyClient _modifyClient() {
    return grpc.PhaseModifyClient(
      GRPCClient().client(),
      options: getCommonCallOptions(),
    );
  }

  Future<List<pb.PhaseSummary>> queryPlanPhaseSummary(int planID) async {
    print("PhaseService.queryPlanPhaseSummary");
    final req = pb.QueryPlanPhaseSummaryRequest.create();
    req.planId = planID;
    final res = await _client().queryPlanPhaseSummary(req);
    return res.phases;
  }

  Future<pb.PhaseSummary> queryPhaseSummary(int planID, int phaseID) async {
    print("PhaseService.queryPhaseSummary");
    final req = pb.QueryPhaseSummaryRequest.create();
    req.planId = planID;
    req.phaseId = phaseID;
    final res = await _client().queryPhaseSummary(req);
    return res.phase;
  }

  Future<void> deletePhase(int phaseId) async {
    print("PhaseService.deletePhase");
    final req = pb.DeletePhaseRequest.create();
    req.phaseId = phaseId;
    await _modifyClient().deletePhase(req);
  }

  Future<int> createPhase(pb.CreatePhaseRequest req) async {
    print("PhaseService.createPhase");
    final res = await _modifyClient().createPhase(req);
    return res.phaseId;
  }

  Future<void> updatePhase(pb.UpdatePhaseRequest req) async {
    print("PhaseService.updatePhase");
    await _modifyClient().updatePhase(req);
  }
}
